### PR TITLE
docs: Update submitting guidelines

### PR DESCRIPTION
**Pull request by nilason**
Updates submitting guidelines, following recent additions of clang-format and pre-commit.

### CODE REVIEW

File: doc/development/submitting/submitting_c.md [hunk 1]
```diff
@@ -344,33 +344,19 @@ Notes:
...
+to be formatted with `clang-format` (currently with version 15+). The most
+convenient method to install clang-format and format files is [using pre-commit](./submitting.md#use-pre-commit).
+
+Alternatively, using separately installed clang-format on modified files:
 
 ```bash
--bad -bap -bbb -br -bli0 -bls -cli0 -ncs -fc1 -hnl -i4
--nbbo -nbc -nbfda -nbfde -ncdb -ncdw -nce -nfca -npcs -nprs
--npsl -nsc -nsob -saf -sai -saw -sbi0 -ss --no-tabs
+clang-format -i <new_or_modified_file.c>
```
Comment (neteler) [inline doc/development/submitting/submitting_c.md:357]: Perhaps add a note from where the `clang-format` settings come? (point to our [file in GH](https://github.com/OSGeo/grass/blob/main/.clang-format) or so?)

Comment (nilason) [reply doc/development/submitting/submitting_c.md:358]: Good point, added.

**Comment by nilason** [comment]
Also added note on how to disable (uninstall) pre-commit.

File: doc/development/submitting/submitting.md [hunk 2]
```diff
@@ -19,6 +19,43 @@ Be sure to check your code against these rules:
...
+
+Pre-commit will then be automatically triggered by git commit command. It is
+also possible to run manually, e.g:
+
+```bash
+pre-commit run clang-format --all-files
+pre-commit run black --all-files
+```
+
+The pre-commmit hooks are defined in
```
Comment (nilason) [inline doc/development/submitting/submitting.md:45]: There is a typo here (pre-commmit). I'll hold on the push for this, in case there are more...

**Comment by nilason** [comment]
I believe this is ready for merge.

File: doc/development/submitting/submitting.md [hunk 3]
```diff
@@ -19,6 +19,43 @@ Be sure to check your code against these rules:
...
+```bash
+cd <grass_source_dir>
+
+# once per repo
+python -m pip install pre-commit
+```
+
+Pre-commit will then be automatically triggered by git commit command. It is
+also possible to run manually, e.g:
+
```
Comment (wenzeslaus) [inline doc/development/submitting/submitting.md:39]: Missing `pre-commit install` somewhere around here.

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
Looks good in general and very good changes overall. Thanks!

File: doc/development/submitting/submitting.md (hunk 3 above)
Comment (nilason) [reply doc/development/submitting/submitting.md:41]: Whoa, you're right! Mixed up with pip install. Fixed.

**Review by wenzeslaus — APPROVED** [review]

**Comment by nilason** [comment]
I added a part to submitting_c.md on using `grass_clang_format.sh`, please check.

**Review by wenzeslaus — APPROVED** [review]
Great. Thanks!

**Comment by wenzeslaus** [comment]
Partially, but only partially, off topic: The pre-commit is not enforced, but it would be a shame to have unsynced rules in pre-commit and in CI. Did you/can you look into `pre-commit run --all-files` for CI (in addition to what we are running now)?

**Comment by nilason** [comment]
> Partially, but only partially, off topic: The pre-commit is not enforced, but it would be a shame to have unsynced rules in pre-commit and in CI. Did you/can you look into `pre-commit run --all-files` for CI (in addition to what we are running now)?

This is the current status:

```
trim trailing whitespace.................................................Passed
fix end of files.........................................................Passed
markdownlint.............................................................Passed
black....................................................................Passed
flake8...................................................................Failed
clang-format.............................................................Passed
yamllint.................................................................Passed
```

Flake8 fails with `--all-files` because in that mode it doesn't respect .flake8 (we could add all exceptions to .pre-commit-config too, then we have to keep the two in sync), works however with ordinary commit, but otherwise present state is good. This after the clang-format PRs, #2785, #2762 and  #2790.

I've been a bit concerned about how to enforce the "trim trailing whitespace", "fix end of files", "markdownlint" and "yamllint", the super-lint settings/abilities isn't/can't be in absolute sync with pre-commit. For example, pre-commit's markdownlint errors with too long lines, whereas super-lint only issues warnings, see eg. the RFC PRs (the present state of them wouldn't pass pre-commit).
I've been looking briefly on [pre-commit-ci](https://pre-commit.ci), which could as part of an action even "add" a commit to PRs applying the "fixing-hooks" like "trim trailing whitespace" or "black", but I'm not quite sure how it works in practice.

**Comment by wenzeslaus** [comment]
I think it's okay when let's say Super-Linter is less strict if pre-commit runs in CI too with a reasonable error message.

The pre-commit doc says:

> pre-commit can also be used as a tool for continuous integration. For instance, adding `pre-commit run --all-files` as a CI step will ensure everything stays in tip-top shape. To check only files which have changed, which may be faster, use something like `pre-commit run --from-ref origin/HEAD --to-ref HEAD`

However, ignoring .flake8 is not good at this point given the amount of exceptions/ignores we need to have now.

**Comment by wenzeslaus** [comment]
...I mean good in the sense that you can't simply run all the pre-commit checks on all files in the CI, but otherwise it works just fine I think.

**Comment by nilason** [comment]
Usually there is no need to run on all files, why would there be? You only check the modified files for commit (staged), no need to run pre-commit explicitly.

**Comment by echoix** [comment]
I'm on the side that if these style checks must be enforced in order to _integrate_ changes to the code, it should be a job for _continuous integration_. And rather than ask everyone to run a specific script to have a certain output, why don't we simply make it on the CI that applies the changes directly to the PR (fixes). 

It was from a discussion in one of the grass repos last year that I learned about Megalinter and Super-Linter, and these "fixes" I know that MegaLinter makes it possible to commit them in the PR, if a linter provides fixes automatically. I have been contributing to MegaLinter since that time I learned about it (this and grass are my two actively followed projects). I also know that at least for the MegaLinter side, it makes it possible to only run on the new/modified files, and Super-Linter might do this too, since they were the same project a couple years ago.

**Comment by nilason** [comment]
Every push to GitHub here on grass is very “costly” in CI time and   the slots easily filled. Therefore I’d say it is should be _recommended_ to pre-empt unnecessary pushes using pre-commit. All the same, “fixable” hooks would be good to make CI take care of, just in case.

**Comment by echoix** [comment]
> Every push to GitHub here on grass is very “costly” in CI time and   the slots easily filled. 

I agree that the CI is quite long, there is a lot of duplicated building, and it's not "fast" either. There can be some work on there to be done if it's really a priority for the project 😃 Correct me if I'm wrong, but the concurrent CI instances are shared between OSGeo repos, right?


> Therefore I’d say it is should be _recommended_ to pre-empt unnecessary pushes using pre-commit. All the same, “fixable” hooks would be good to make CI take care of, just in case.

Completely agree with both of these points

**Comment by wenzeslaus** [comment]
Just to be clear, this can be merged, but this discussion is good to have!

> ...the concurrent CI instances are shared between OSGeo repos...

Yes, I think it is per organization although I don't see that explicitly stated in the docs.

> ...a lot of time is spend on waiting to run the checks.

That's most of the time on a "bad" day (bad meaning very active, so good :-)). The runs are all under 2 hours, most under 20 minutes. That's not to say that faster or less jobs would not be better. Also, @echoix if you want to add MegaLinter or replace Super-Linter, go for it. If I recall correctly, the only reason we have Super-Linter and not MegaLinter is that MegaLinter did not exist yet.

> > Therefore I’d say it is should be _recommended_ to pre-empt unnecessary pushes using pre-commit.

...or editor settings. For example, some QtCreator plugin can apply clang-format on file save. However, pre-commit is much easier to set up than many different checks.

> All the same, “fixable” hooks would be good to make CI take care of, just in case.

I'm little hesitant about that because it makes it harder for people not familiar with Git - something else is changing their branch and they need to update. I'm using it in https://github.com/ncsu-geoforall-lab/grass-gis-on-hpc-henry2/ and I have mixed feelings about it mostly because I was not able to configure my local tools same as MegaLinter.

BTW, this is on the first Google page of "pre-commit enforced in ci":

> Pre-commit hooks are a waste of time.
>
> Why? Because they enforce standards too early. Pre-commit hooks make us wait for a quality check on every commit.

I don't necessarily agree (will see in some time), just something to be aware of.

**Comment by nilason** [comment]
> Just to be clear, this can be merged, but this discussion is good to have!

Figured that much :-) , done!

> I'm little hesitant about that because it makes it harder for people not familiar with Git - something else is changing their branch and they need to update.

I agree and that is also something I have been considering on the con side of CI commit hooks.

> BTW, this is on the first Google page of "pre-commit enforced in ci":

> > Pre-commit hooks are a waste of time.

> > Why? Because they enforce standards too early. Pre-commit hooks make us wait for a quality check on every commit.

I think this more reflects the use of pre-commit locally, not on CIs. Ideally a "pre-push" hook, triggered on `git push` would perhaps be less disturbing than on `git commit`, but I'm not aware of such hook.

**Comment by wenzeslaus** [comment]
> I think this more reflects the use of pre-commit locally, not on CIs.

Yes, I think the point there was CI time is still cheaper than developer time. (Plus you can't use pre-commit for enforcement anyway.) Our pre-commit is fast right now, so no trouble there.

> Ideally a "pre-push" hook, triggered on `git push` would perhaps be less disturbing than on `git commit`, but I'm not aware of such hook.

I did not investigate, but pre-commit doc mentions push and this [SO](https://stackoverflow.com/a/14672883/574907) refers to the Git feature.

**Pull request merged**